Apollo koa integration

HriBB opened the pull request:
[KoaJS](https://github.com/koajs) apollo-server integration. All tests pass, but I still need to test it in a real-life application.

TODO:

- [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
- [ ] Make sure all of the significant new logic is covered by tests
- [ ] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

helfer commented:
@hribb First of all, thanks a lot for the PR!

It looks like all the tests pass, but there are still some linter errors that you should fix:

```
src/integrations/expressApollo.test.ts[28, 26]: block is empty
src/integrations/hapiApollo.test.ts[31, 26]: block is empty
src/integrations/koaApollo.test.ts[16, 10]: Unused import: 'OperationStore'
src/integrations/koaApollo.ts[2, 13]: Unused import: 'koaBody'
```

I'll add comments about the code inline.

helfer commented on src/integrations/integrations.test.ts:
```diff
@@ -80,6 +84,7 @@ export default (createApp: CreateAppFunc) => {
                   query: 'query test{ testString }',
               });
           return req.then((res) => {
+              destroyApp(app);
```
Did `afterEach` not work? That would be preferable to adding the call explicitly in every test, I think.

helfer commented on src/integrations/koaApollo.test.ts:
```diff
@@ -0,0 +1,66 @@
+}
+
+describe('graphqlHTTP', () => {
+   it('returns express middleware', () => {
```
I think this test isn't necessary any more. We could also remove it from the express tests.

helfer commented on src/integrations/koaApollo.test.ts:
```diff
@@ -0,0 +1,66 @@
+});
+
+describe('renderGraphiQL', () => {
+  it('returns express middleware', () => {
```
Like the previous one, this is also not necessary. But if we keep it, it should be renamed.

helfer commented on typings.json:
```diff
@@ -6,10 +6,13 @@
   },
   "globalDependencies": {
     "body-parser": "registry:dt/body-parser#0.0.0+20160619023215",
+    "cookies": "registry:dt/cookies#0.5.1+20160316171810",
```
Where do we use this? I can't find any reference in the code. Is this something that Koa requires?

HriBB replied:
I guess this is needed by koa typings. If I remove it, I get this error
```
import * as cookies from "cookies";

typings/globals/koa/index.d.ts(5,30): error TS2307: Cannot find module 'cookies'.
```
I am a typescript beginner so I might be doing something wrong. @nnance can you please check this?

helfer commented on src/integrations/koaApollo.test.ts:
```diff
@@ -0,0 +1,66 @@
+// tslint:disable-next-line
+import * as koa from 'koa';
+import * as koaRouter from 'koa-router';
+import * as koaBody from 'koa-bodyparser';
```
This import isn't used according to tslint.

helfer commented on src/integrations/koaApollo.test.ts:
```diff
@@ -0,0 +1,66 @@
+import ApolloOptions from './apolloOptions';
+import { apolloKoa, graphiqlKoa } from './koaApollo';
+
+import { OperationStore } from '../modules/operationStore';
```
This import isn't used.

helfer commented on src/integrations/koaApollo.ts:
```diff
@@ -0,0 +1,130 @@
+import * as koa from 'koa';
+import * as koaBody from 'koa-bodyparser';
```
Unused import according to tslint.

helfer commented on src/integrations/expressApollo.test.ts:
```diff
@@ -25,6 +25,8 @@ function createApp(options: CreateAppOptions = {}) {
   return app;
 }
 
+function destroyApp(app) {}
```
I think you should get rid of this and instead make the `destroyApp` argument optional. That way you only have to modify one existing test file.

HriBB commented:
> is this a good plan to introduce the alpha version of koa? will this alienate people who are using 1.x version of koa?

You are right, it's not a good plan, but I am not sure how to support both versions. Problem is that v2 uses a different syntax with async/await, where v1 uses generators/yield.

koa@1.x
```
app.use(function *(){
  this.body = 'Hello World';
});
```
koa@2.x
```
app.use(function(ctx){
  ctx.body = 'Hello World';
});
```
Not really sure how to tackle this. We could simply export two different versions such as `apolloKoa` and `apolloKoa2`. Not the best solution, but it could work, if there is a way to install both versions at once with npm. Is there?

HriBB commented:
@helfer and @nnance I removed all unnecessary tests, fixed linter issues, made `destroyApp()` optional and moved it into `afterEach()` mocha hook. Please check it out and let me know what you think.

helfer commented:
@HriBB thanks, that looks great! 

Regarding the use of Koa 2 vs. 1, I found this interesting Medium post: https://medium.com/@l1ambda/why-you-should-use-koa-with-node-js-7c231a8174fa#.8h6opnbvy

The tldr is that 2.0 alpha is considered stable, but they won't officially recommend it until Node has native support for async/await and Koa 2.0 can be used without Babel.

I'd say we land this PR as it is, and consider rewriting it for Koa 1 if there are many requests for it.

helfer commented:
@HriBB can you update the Changelog to mention Koa support? The format should be something along the lines of

* added Koa integration (@HriBB in #59)

HriBB commented:
@helfer done!

HriBB commented:
I hope I did not break anything with merging upstream ...

helfer commented:
Perfect, merged!
